tests/Feature/CountryControllerTest.php

```diff
@@ -0,0 +1,189 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Tests\Feature;
+
+use App\Models\Country;
+use Generator;
+use Illuminate\Foundation\Testing\RefreshDatabase;
+use Inertia\Testing\AssertableInertia as Assert;
+use Tests\TestCase;
+
+class CountryControllerTest extends TestCase
+{
+    use RefreshDatabase;
+
+    public function testCountryIndex(): void
+    {
+        Country::factory()->count(3)->create();
+
+        $this->get("/countries")->assertInertia(
+            fn(Assert $page) => $page
+                ->component(value: "Countries")
+                ->has("countries.data", 3),
+        );
+    }
+
+    public function testCountryStore(): void
+    {
+        $country = [
+            "name" => "Poland",
+            "latitude" => 44.543,
+            "longitude" => -43.122,
+            "iso" => "pl",
+        ];
+
+        $this->post(route(name:"countries.store"), data: $country);
+
+        $this->assertDatabaseHas(table:"countries", data: $country);
+    }
+
+    public function testCountryStoreIsRefusingDuplicate(): void
+    {
+        $country = [
+            "name" => "Poland",
+            "latitude" => 44.543,
+            "longitude" => -43.122,
+            "iso" => "pl",
+        ];
+
+        Country::query()->create([
+            "name" => $country["name"],
+            "latitude" => -55.54323,
+            "longitude" => 42.3721,
+            "iso" => $country["iso"],
+        ])->toArray();
+
+        $this->post(route(name:"countries.store"), data: $country);
+
+        $this->assertDatabaseMissing(table: "countries", data: $country);
+    }
+
+    /**
+     * @dataProvider invalidCountryDataProvider
+     */
+    public function testCountryCannotBeCreatedWithInvalidData(array $data, array $expectedErrors): void
+    {
+        $response = $this->post(route("countries.store"), $data);
+
+        $response->assertSessionHasErrors($expectedErrors);
+    }
+
+    /**
+     * @dataProvider invalidCountryDataProvider
+     */
+    public function testCountryCannotBeUpdatedWithInvalidData(array $data, array $expectedErrors): void
+    {
+        $country = Country::factory()->create();
+
+        $response = $this->patch(route("countries.update", ["country" => $country]), $data);
+
+        $response->assertSessionHasErrors($expectedErrors);
+    }
+
+    public static function invalidCountryDataProvider(): Generator
+    {
+        yield "country with empty credentials is invalid" => [
```

Comment:
Remove `is invalid` in all yields.
These strings describe test steps, so this `is invalid` is not needed.